content/ADR-146-unity-2021-material-upgrade.md

```diff
@@ -0,0 +1,77 @@
+---
+layout: adr
+adr: 146
+title: Upgrade AB Materials to Unity 2021
+date: 2022-12-02
+status: Live
+type: ADR
+spdx-license: CC0-1.0
+slug: /adr/TEMPLATE
+authors:
+- ajimenez
+---
+
+## Abstract
+An algorithm to manually upgrade materials packed in Asset Bundles made in Unity 2020 is required when using Unity 2021. This document will go into the details of the SRP Batcher optimizations and how to navigate the issues in URP serialization. 
+
+## Problem
+All our Asset Bundles (by the time this ADR is written) are converted using Unity 2020 version. Although the Asset Bundles serialization between both versions have not changed, the URP version has. That means we can reuse the asset bundles but not the materials without manually upgrading them.
+
+## Glossary
+- AB (aka [Asset Bundle](https://docs.unity3d.com/Manual/AssetBundlesIntro.html)): A proprietary format used by unity to store assets.
+- [SRP Batcher](https://docs.unity3d.com/Manual/SRPBatcher.html): A draw call optimization based on reusing the same shader variant.
+- Shader Variant: A Shader with an specific set of keywords and defines.
+
+## Solution Space Exploration
+
+- ❌ **Reconvert the world**: The idea is simple, use a Unity 2021 AB converter and redeploy all the ABs, assuming we could create a cron job that would spend a few days converting every scene in the world and every wearable to use the new version. Even with that, we'd need some new infrastructure on top of the systems such as:
+  - Upgrading the AB converter: Not trivial, we need to solve some issues with multithreading first (the AB Converter is stuck in an older commit).
+  - Versioning: The renderer must understand if it can consume an AB or is not prepared for it's version.
+  - Content size: Literally duplicating the size of the whole world to allocate new versions of ABs is something we should do if nothing else work.
+- ✔ **Manual upgrade**: Ideally we should be able to gather all the information from the material itself and create a new one with the properties well defined. 
+
+## Upgrading materials from 2020 to 2021
+Unfortunately Unity is using the same class to serialize materials from Unity 2020 to Unity 2021, that means that we have no access to anything the new material doesnt provide. That makes finding stuff like the rendering target (Opaque, Alpha Test, Transparency) challenging because the new URP changed the way keywords are serialized. 
```

Comment:
Small suggestion:
```suggestion
Unfortunately, Unity is using the same class to serialize materials from Unity 2020 to Unity 2021, which means that we have no access to anything the new material doesn't provide. That makes finding stuff like the rendering target (Opaque, Alpha Test, Transparency) challenging because the new URP changed the way keywords are serialized. 
```